[DPE-9356] Update PostgreSQL mount paths to /var/lib/pg with 16/main subdirectory

marceloneppel opened the pull request:
## Issue

The database uses the mount points provided by Juju as the data directories. This fails the database bootstrap process when there is some content in those mount points, like `lost+found`.

Also, the storage mount paths use `/var/lib/postgresql/*`, which doesn't follow the `16/main` subdirectory convention expected by PostgreSQL tooling.

## Solution

Following the spec [DA230](https://docs.google.com/document/d/1eCVf0OH0BZ6GAt9bhxhgObWubPWttn5IZa0GXVtdvJo/edit?tab=t.0), migrate storage mount paths from `/var/lib/postgresql/*` to `/var/lib/pg/*` with a `16/main` subdirectory structure, and create a compatibility symlink from the standard PostgreSQL data path.

### Path changes

| Storage | Old path | New path |
|---------|----------|----------|
| Archive | `/var/lib/postgresql/archive` | `/var/lib/pg/archive/16/main` |
| Data | `/var/lib/postgresql/data/pgdata` | `/var/lib/pg/data/16/main` |
| Logs (WAL) | `/var/lib/postgresql/logs` | `/var/lib/pg/logs/16/main/pg_wal` |
| Temp | `/var/lib/postgresql/temp` | `/var/lib/pg/temp/16/main/pgsql_tmp` |

### Key changes

- **Storage mounts** (`metadata.yaml`): Update all four storage mount locations to `/var/lib/pg/*`
- **Named constants** (`constants.py`): Extract storage paths into `ARCHIVE_STORAGE_PATH`, `DATA_STORAGE_PATH`, `LOGS_STORAGE_PATH`, `TEMP_STORAGE_PATH` and use them consistently across the codebase
- **Symlink compatibility**: Create a symlink from `/var/lib/postgresql/16/main` → `/var/lib/pg/data/16/main` so PostgreSQL and Patroni reference the canonical path; always recreate the symlink (removing any real directory the OCI image ships first)
- **Patroni & pgBackRest templates**: Update `patroni.yml.j2` and `pgbackrest.conf.j2` with new `data_dir`, `waldir`, and temp tablespace paths
- **Backup/restore** (`backups.py`): Clear all storage directories during restore preparation so `pg_basebackup --waldir`/`--tablespace` options work on replicas; use actual pgdata path (not symlink) for file operations
- **Async replication** (`async_replication.py`): Prevent system ID mismatch by only clearing pgdata on non-leaders after the standby leader is confirmed running; check `member_started` before clearing pgdata on standby units; fix disk usage check to use the actual pgdata path
- **Replica startup** (`charm.py`): Defer replica startup until its endpoint is present in the members list, preventing `pg_basebackup` rejections; clear stale WAL/temp directories repeatedly until pgdata is fully populated (`PG_VERSION` exists)
- **Security**: Fix shell injection risk by passing command arguments as lists instead of split strings
- **Upgrade tests disabled**: Temporarily disable upgrade-related spread tests since upgrades cannot be performed when storage definitions change
- **New integration test**: Add `test_symlinks.py` to verify the data directory symlink works correctly

## Checklist
- [x] I have added or updated any relevant documentation.
- [ ] I have cleaned any remaining cloud resources from my accounts.

Fixes: https://github.com/canonical/postgresql-k8s-operator/issues/1009

marceloneppel committed:
Update PostgreSQL mount paths to /var/lib/pg with 16/main subdirectory

Migrate storage mount paths from /var/lib/postgresql/* to /var/lib/pg/*
and configure PostgreSQL 16 data directories to use the 16/main
subfolder structure.

Path changes:
- Archive: /var/lib/postgresql/archive → /var/lib/pg/archive
- Data:    /var/lib/postgresql/data/pgdata → /var/lib/pg/data/16/main
- Logs:    /var/lib/postgresql/logs → /var/lib/pg/logs/16/main (WAL)
- Temp:    /var/lib/postgresql/temp → /var/lib/pg/temp/16/main

Key changes:
- Update metadata.yaml storage mount locations
- Update POSTGRESQL_DATA_PATH constant
- Configure pgdata_path, waldir, and temp tablespace paths
- Add make_parents=True to Pebble make_dir for nested directory creation
- Update pgbackrest.conf.j2 with new data path
- Update patroni.yml.j2 with new data_dir and waldir
- Update async_replication.py cleanup paths
- Update authorisation_rules_observer.py Patroni config path

Note: Archive path (/var/lib/pg/archive) intentionally does not use the
16/main subfolder as it is not used in Patroni or PostgreSQL configuration.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix(charm): robustly patch OCI health-check script on pebble-ready/_fix_pod

Patch the OCI image health-check script (/scripts/self-signed-checker.py) to use the
charm's storage path for the peer CA file instead of the hardcoded
/var/lib/postgresql/data/peer_ca.pem. Make the operation idempotent and
non-blocking, and call it from pebble-ready and early pod-fix flows to avoid
race conditions with Pebble layer installation.

Add _fix_health_check_script(container) to read, replace, and write the script.
Log and handle pull/push errors; on push failure attempt an in-container sed
fallback (exec ["sh","-c","sed -i 's|OLD|NEW|g' ..."]).
Invoke the fixer from _on_postgresql_pebble_ready (after _create_pgdata) and
from _fix_pod (early) so the script is corrected as early as possible.
Update unit tests to cover the patching behaviour and the exec fallback.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
chore: fix linter warnings and apply ruff formatting

- Replace broad try/except/pass patterns with contextlib.suppress where appropriate
  (fixes SIM105 and S110 warnings).
- Ensure the _fix_health_check_script fallback exec wait is wrapped with
  contextlib.suppress(Exception) to avoid swallowing exceptions without intent.
- Use contextlib.suppress to defensively guard _fix_pod health-check fixes while
  still logging inner failures.
- Re-order and tidy imports to satisfy ruff import ordering rules.
- Apply ruff format across src/, tests/, and scripts/ to normalize formatting.

Notes:
- These are non-functional changes (linting / formatting); behavior unchanged.
- Formatting touched multiple test and helper files; no logic was altered.
- Tests were not executed as part of this change; please run the unit/integration
  suites locally (tox -e unit / tox -e integration) before pushing if desired.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
chore(tests): apply ruff formatting to integration tests

- Apply ruff/auto-formatting to integration test files to satisfy lint/format checks:
  - tests/integration/ha_tests/conftest.py
  - tests/integration/ha_tests/helpers.py
  - tests/integration/test_invalid_database.py
  - tests/integration/test_invalid_extra_user_roles.py

- Non-functional: only whitespace/line-wrapping/formatting changes were made.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
test(spread): disable upgrade integration tests

Temporarily disable the upgrade-related integration tests by constraining
their spread task configurations so they will not be scheduled in CI.

Changes:
- Added a `systems:` block to the following spread task files to prevent
  these tasks from running across the broader CI fleet:
  - tests/spread/test_upgrade.py/task.yaml
  - tests/spread/test_upgrade_skip_pre_upgrade_check.py/task.yaml
  - tests/spread/test_upgrade_from_stable.py/task.yaml
  - tests/spread/test_async_replication_upgrade.py/task.yaml

Reason:
- Upgrades cannot be performed when storage definitions change. These
  tests fail in situations where storage definitions are modified, so
  they are being disabled until the upgrade path for storage changes is
  resolved.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
tests: add PGDATA_PATH and use in integration tests

- Add PGDATA_PATH = f"{STORAGE_PATH}/16/main"
- Update tests/integration/test_storage.py and tests/integration/test_charm.py to use PGDATA_PATH
- Fix import ordering and lint issues

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel commented on src/charm.py:
```diff
@@ -1153,14 +1181,86 @@ def _create_pgdata(self, container: Container):
+            temp_path,
         ]).wait()
 
+    def _fix_health_check_script(self, container: Container) -> None:
```
This needs to be removed and instead we need to update the script in it's repository (https://github.com/canonical/charmed-postgresql-rock/blob/4dbbaa70461ee8977244ad0f322b274a600caa53/local/scripts/self-signed-checker.py).

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Create data directory symlink (#1217)

* Update canonical/data-platform-workflows action to v41 (#1211)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

* feat(storage): add symlink from default PostgreSQL data path

Create a symlink from /var/lib/postgresql/16/main to /var/lib/pg/data/16/main
to provide compatibility with the standard PostgreSQL data directory location.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

---------

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>
Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

marceloneppel committed:
Merge branch 'change-storage-mounts' of github.com:canonical/postgresql-k8s-operator into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix: use consistent symlink path for PostgreSQL data directory

Update POSTGRESQL_DATA_PATH and pgdata_path to use the symlink
path (/var/lib/postgresql/16/main) instead of the actual storage
path (/var/lib/pg/data/16/main) for Patroni, pgbackrest, and tests.

This ensures all components reference the canonical symlink path
that PostgreSQL and Patroni expect.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix(async-replication): prevent system ID mismatch for standby replicas

Non-leader units in a standby cluster now clear pgdata only after the
standby leader is confirmed running, preventing system ID mismatch errors
during pg_basebackup. Also fixes disk usage check to use actual pgdata
path and handles Patroni API unavailability gracefully during slot sync.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
style: improve code formatting in patroni and async_replication modules

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix(backups): use actual pgdata path instead of symlink for backup operations

Update pgbackrest config template to use the proper POSTGRESQL_DATA_PATH
constant and fix _empty_data_files to remove the actual directory contents
rather than just the symlink, ensuring backup and restore operations work
correctly.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix(tests): update clean-data-dir.sh paths and use safer deletion

- Update data directory paths to match current structure (/var/lib/pg/data/16/main, etc.)
- Replace rm -rf with find -delete to preserve mount point directories

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix(backup): clear all storage directories during restore for replica compatibility

Clear archive, data, logs, and temp directories during backup restore
preparation. This ensures pg_basebackup can use --waldir and --tablespace
options when new replicas join after a restore, as these options require
empty directories.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
refactor(charm): remove stale storage cleanup for replica initialization

Remove the logic that cleared logs and temp directories when a non-leader
unit rejoined with an empty data directory. This cleanup was intended for
pg_basebackup during restore operations but is no longer needed.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix(charm): clear stale storage dirs before replica initialization

Clear WAL and tablespace directories when a replica joins an initialized
cluster, as PersistentVolumes may retain data from previous pods and
pg_basebackup requires empty target directories.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix(async-replication): check member_started before clearing pgdata on standby

When a new unit joins an existing standby cluster, check if the database
is already running before clearing pgdata. Previously, _clear_pgdata()
was called unconditionally before the member_started check, which could
delete the data directory while PostgreSQL was running - causing it to
crash and leaving the unit stuck in "awaiting for member to start".

Now the order is:
1. If member is already started, return early (unit is working)
2. If member is not started, clear pgdata if needed and continue setup

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix: remove health check script patching workaround

The `_fix_health_check_script` method was a workaround for an incorrect
hardcoded certificate path in the OCI image. This is no longer needed
as the upstream image has been fixed. Update the OCI image reference
accordingly.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix(storage): ensure pgdata symlink is always recreated and stale dirs are cleared on retry

The OCI image ships /var/lib/postgresql/16/main as a real directory,
which prevents the symlink from being created on replicas. Always
recreate the symlink (removing the real directory first if needed)
instead of skipping when the path already exists.

Also change stale storage directory cleanup to repeat until pgdata is
fully populated (PG_VERSION exists), rather than running only once per
unit lifecycle. This ensures retries after a failed pg_basebackup find
clean directories.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix: prevent pg_wal symlink breakage on replica startup

Use exact pg_wal and pgsql_tmp subdirectories for --waldir and temp
tablespace paths instead of their parent directories. Also defer
replica startup until its endpoint is present in the members list,
preventing pg_basebackup rejections that trigger remove_data_directory()
calls racing with _create_pgdata().

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
refactor: replace hardcoded storage paths with named constants

Extract storage mount paths (/var/lib/pg/{archive,data,logs,temp}) into
named constants in constants.py and use them consistently across
backups, async replication, Patroni config rendering, and charm logic.
Also fix shell injection risk by passing command arguments as lists
instead of split strings, guard rotate_logs service start behind a
service existence check, and log failed replication slot syncs instead
of silently ignoring the error.

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
docs: fix broken Juju 3.0 release notes URLs

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
fix(metadata): update postgresql-image upstream source to canonical GHCR

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge remote-tracking branch 'origin/16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

carlcsaposs-canonical commented on metadata.yaml:
```diff
@@ -29,19 +29,19 @@ containers:
     type: oci-image
     description: OCI image for PostgreSQL
-    upstream-source: ghcr.io/canonical/charmed-postgresql@sha256:fcb648db2c418403fbf5f58f04892c4677f113526097ca7314cbd9f527bb1baf  # renovate: oci-image tag: 16.11-24.04_edge
+    upstream-source: ghcr.io/canonical/charmed-postgresql@sha256:7ee3f31ce88a48a94634d852a50e6cf3c50c6773eb885e4b896d3039150aa633
```
any reason to remove renovate comment?

marceloneppel replied:
Fixed on https://github.com/canonical/postgresql-k8s-operator/pull/1186/changes/6689a53216777bac257d5bb9506c09376bd0aa1b.

carlcsaposs-canonical reviewed (dismissed):
> Clear all storage directories during restore preparation

> clearing pgdata on non-leaders after the standby leader is confirmed running

I remember awhile ago Mykola mentioning that we should consider forbidding restore over existing data to avoid data loss and instead requiring restore to a new deployment/empty storage. I think this happened after the incident where in production a backup was restored to a unit we didn't realize was primary at one point, causing data loss since only that unit had the most recent data

Not sure if it's feasible to consider now, but was wondering if we should delete data during restore

marceloneppel commented on src/backups.py:
```diff
@@ -272,15 +276,24 @@ def _create_bucket_if_not_exists(self) -> None:
-                raise
+        # Clear all storage directories, not just data. The logs directory must be cleared
+        # so that when new replicas join after restore, pg_basebackup can use the --waldir
+        # option (which requires an empty directory).
```
Yes, it only contains WAL nowadays. Mohamed wants to in the future move the database logs that are currently at  `/var/log/postgresql`, when we'll need to ensure to not delete those logs.

marceloneppel commented:
> I remember awhile ago Mykola mentioning that we should consider forbidding restore over existing data to avoid data loss and instead requiring restore to a new deployment/empty storage. I think this happened after the incident where in production a backup was restored to a unit we didn't realize was primary at one point, causing data loss since only that unit had the most recent data
Not sure if it's feasible to consider now, but was wondering if we should delete data during restore

Very good point. You're absolutely right about that, @carlcsaposs-canonical. Let me extract this point to another task so I don't forget about this. I remember there was a discussion about this, which also extends to the async replication standby cluster data deletion.

marceloneppel committed:
ci: restore renovate hint for postgresql oci-image resource

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel commented:
Create https://github.com/canonical/postgresql-k8s-operator/issues/1262 for the above issue.

marceloneppel commented on metadata.yaml:
```diff
@@ -29,19 +29,19 @@ containers:
     type: oci-image
     description: OCI image for PostgreSQL
-    upstream-source: ghcr.io/canonical/charmed-postgresql@sha256:fcb648db2c418403fbf5f58f04892c4677f113526097ca7314cbd9f527bb1baf  # renovate: oci-image tag: 16.11-24.04_edge
+    upstream-source: ghcr.io/canonical/charmed-postgresql@sha256:7ee3f31ce88a48a94634d852a50e6cf3c50c6773eb885e4b896d3039150aa633  # renovate: oci-image tag: 16.11-24.04_edge
```
Thanks!

marceloneppel committed:
Update OCI image hash

Co-authored-by: Dragomir Penev <6687393+dragomirp@users.noreply.github.com>
Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@gmail.com>

marceloneppel committed:
Merge branch '16/edge' into change-storage-mounts

taurus-forever reviewed (dismissed):
Exceptionally well done! Nothing found after incentive testing. :partying_face: 

One question: PR description says `/var/lib/pg/archive/16/main`
but the deployed system has no 16/main for archive:
```
root@postgresql-k8s-3:/var/lib/postgresql# ls -la /var/lib/pg/archive/                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
total 24                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
drwxr-xr-x 3 postgres postgres  4096 Mar  2 23:31 .                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           
drwxr-xr-x 6 root     root      4096 Mar  2 23:32 ..                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
drwx------ 2 root     root     16384 Mar  2 23:31 lost+found  
```

Outdated PR description? @marceloneppel

marceloneppel committed:
Merge branch '16/edge' into change-storage-mounts

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
Merge branch '16/edge' into change-storage-mounts

marceloneppel committed:
test: update integration tests to use CHARM_BASE_NOBLE

Replace CHARM_BASE references with CHARM_BASE_NOBLE in HA self-healing
and new relations integration tests

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>

marceloneppel committed:
test: update HA tests to use CHARM_BASE_NOBLE and latest/edge channel

Signed-off-by: Marcelo Henrique Neppel <marcelo.neppel@canonical.com>
